# Deep Learning Model Life-Cycle

A model has a life-cycle, and this very simple knowledge provides the backbone for both modeling a dataset and understanding the tf.keras API.

The five steps in the life-cycle are as follows:

1. Define the model.
2. Compile the model.
3. Fit the model.
4. Evaluate the model.
5. Make predictions.

# MLP for Binary Classification

In [16]:
# MLP for Binary Classification
from pandas import read_csv
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from tensorflow.keras import Sequential
from tensorflow.keras.layers import Dense
# load the dataset
path = 'https://raw.githubusercontent.com/jbrownlee/Datasets/master/ionosphere.csv'
df = read_csv(path, header=None)
# split into input and output columns
X, y = df.values[:, :-1], df.values[:, -1]
# ensure all data are floating point values
X = X.astype('float32')
# encode strings to integer
y = LabelEncoder().fit_transform(y)
# split into train and test datasets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33)
print(X_train.shape, X_test.shape, y_train.shape, y_test.shape)
# determine the number of input features
n_features = X_train.shape[1]
#print(n_features)
# define model****************************************************************************************
model = Sequential()
model.add(Dense(10, activation='relu', kernel_initializer='he_normal', input_shape=(n_features,)))
model.add(Dense(8, activation='relu', kernel_initializer='he_normal'))
model.add(Dense(1, activation='sigmoid'))
# compile the model***********************************************************************************
model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
# fit the model***************************************************************************************
model.fit(X_train, y_train, epochs=150, batch_size=32, verbose=0)
# evaluate the model**********************************************************************************
loss, acc = model.evaluate(X_test, y_test, verbose=0)
print('Test Accuracy: %.3f' % acc)
#print(loss)
# make a prediction***********************************************************************************
row = [1,0,0.99539,-0.05889,0.85243,0.02306,0.83398,-0.37708,1,0.03760,0.85243,-0.17755,0.59755,-0.44945,0.60536,-0.38223,0.84356,-0.38542,0.58212,-0.32192,0.56971,-0.29674,0.36946,-0.47357,0.56811,-0.51171,0.41078,-0.46168,0.21266,-0.34090,0.42267,-0.54487,0.18641,-0.45300]
#row = [0,0,	0,	0,	0,	0	,1,	-1,	0	,0,	-1,	-1,	0,	0,	0,	0	,1,	1,	-1,	-1,	0	,0	,0	,0	,1,	1,	1,	1,	0,	0,	1,	1,	0	,0]
yhat = model.predict([row])
print('Predicted: %.3f' % yhat)


(235, 34) (116, 34) (235,) (116,)
Test Accuracy: 0.948
Predicted: 0.984


In this case above, we can see that the model achieved a classification accuracy of 94 percent and then predicted a probability of 0.984 that the one row of data belongs to class 1.

In [17]:
# evaluate the model
loss, acc = model.evaluate(X_test, y_test, verbose=0)
print('Test Accuracy: %.3f' % acc)

# make another prediction
row = [0,0,0,0,0,0,1,-1,0,0,-1,-1,0,0,0,0,1,1,-1,-1,0,0,0,0,1,1,1,1,0,0,1,1,0,0]
yhat = model.predict([row])
print('Predicted: %.3f' % yhat)

Test Accuracy: 0.948
Predicted: 0.001


In this case above, we can see that the model achieved a classification accuracy of 94 percent and then predicted a probability of 0.001 that the one row of data belongs to class 0.

# MLP for Multiclass Classification

In [29]:
# mlp for multiclass classification
from numpy import argmax
from pandas import read_csv
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from tensorflow.keras import Sequential
from tensorflow.keras.layers import Dense
# load the dataset
path = 'https://raw.githubusercontent.com/jbrownlee/Datasets/master/iris.csv'
df = read_csv(path, header=None)
# split into input and output columns
X, y = df.values[:, :-1], df.values[:, -1]
# ensure all data are floating point values
X = X.astype('float32')
# encode strings to integer
y = LabelEncoder().fit_transform(y)
#print(y)
# split into train and test datasets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33)
print(X_train.shape, X_test.shape, y_train.shape, y_test.shape)
# determine the number of input features
n_features = X_train.shape[1]
# define model****************************************************************************************
model = Sequential()
model.add(Dense(10, activation='relu', kernel_initializer='he_normal', input_shape=(n_features,)))
model.add(Dense(8, activation='relu', kernel_initializer='he_normal'))
model.add(Dense(3, activation='softmax'))
# compile the model***********************************************************************************
model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
# fit the model***************************************************************************************
model.fit(X_train, y_train, epochs=150, batch_size=32, verbose=0)
# evaluate the model**********************************************************************************
loss, acc = model.evaluate(X_test, y_test, verbose=0)
print('Test Accuracy: %.3f' % acc)
# make a prediction***********************************************************************************
row = [5.1,3.5,1.4,0.2]
#row = [6.9,3.2,5.7,2.3]
#row = [5.7,2.8,4.5,1.3]
yhat = model.predict([row])
print('Predicted: %s (class=%d)' % (yhat, argmax(yhat)))


(100, 4) (50, 4) (100,) (50,)
Test Accuracy: 0.960
Predicted: [[9.991980e-01 6.892086e-04 1.127535e-04]] (class=0)


In [ ]:
""" In this case above, we can see that the model achieved a classification accuracy of about 96 percent and 
then predicted a probability of a row of data belonging to each class, although class 0 has the highest 
probability.

(100, 4) (50, 4) (100,) (50,)
Test Accuracy: 0.960
Predicted: [[9.991980e-01 6.892086e-04 1.127535e-04]] (class=0) """

In [30]:
# evaluate the model
loss, acc = model.evaluate(X_test, y_test, verbose=0)
print('Test Accuracy: %.3f' % acc)

# make second prediction
row = [6.9,3.2,5.7,2.3]
yhat = model.predict([row])
print('Predicted: %s (class=%d)' % (yhat, argmax(yhat)))

Test Accuracy: 0.960
Predicted: [[0.00249702 0.20922099 0.78828204]] (class=2)


In [ ]:
""" In this case above, we can see that the model achieved a classification accuracy of about 96 percent and 
then predicted a probability of a row of data belonging to each class, although class 2 has the highest 
probability.

(100, 4) (50, 4) (100,) (50,)
Test Accuracy: 0.960
Predicted: [[0.00249702 0.20922099 0.78828204]] (class=2) """

In [31]:
# evaluate the model
loss, acc = model.evaluate(X_test, y_test, verbose=0)
print('Test Accuracy: %.3f' % acc)

# make third prediction
row = [5.7,2.8,4.5,1.3]
yhat = model.predict([row])
print('Predicted: %s (class=%d)' % (yhat, argmax(yhat)))

Test Accuracy: 0.960
Predicted: [[0.01029302 0.53254884 0.45715812]] (class=1)


In [ ]:
""" In this case above, we can see that the model achieved a classification accuracy of about 96 percent and 
then predicted a probability of a row of data belonging to each class, although class 1 has the highest 
probability.

(100, 4) (50, 4) (100,) (50,)
Test Accuracy: 0.960
Predicted: [[0.01029302 0.53254884 0.45715812]] (class=1) """

# MLP for Regression

In [42]:
# mlp for regression
from numpy import sqrt
from pandas import read_csv
from sklearn.model_selection import train_test_split
from tensorflow.keras import Sequential
from tensorflow.keras.layers import Dense
# load the dataset
path = 'https://raw.githubusercontent.com/jbrownlee/Datasets/master/housing.csv'
df = read_csv(path, header=None)
# split into input and output columns
X, y = df.values[:, :-1], df.values[:, -1]
# split into train and test datasets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33)
print(X_train.shape, X_test.shape, y_train.shape, y_test.shape)
# determine the number of input features
n_features = X_train.shape[1]
#print(n_features)
# define model****************************************************************************************
model = Sequential()
model.add(Dense(10, activation='relu', kernel_initializer='he_normal', input_shape=(n_features,)))
model.add(Dense(8, activation='relu', kernel_initializer='he_normal'))
model.add(Dense(1))
# compile the model***********************************************************************************
model.compile(optimizer='adam', loss='mse')
# fit the model***************************************************************************************
model.fit(X_train, y_train, epochs=150, batch_size=32, verbose=0)
# evaluate the model**********************************************************************************
error = model.evaluate(X_test, y_test, verbose=0)
print('MSE: %.3f, RMSE: %.3f' % (error, sqrt(error)))
# make a prediction***********************************************************************************
row = [0.00632,18.00,2.310,0,0.5380,6.5750,65.20,4.0900,1,296.0,15.30,396.90,4.98]
#row = [0.17004,12.5,7.87,0,0.524,6.004,85.9,6.5921,5,311,15.2,386.71,17.1]

yhat = model.predict([row])
print('Predicted: %.3f' % yhat)

(339, 13) (167, 13) (339,) (167,)
MSE: 48.764, RMSE: 6.983
Predicted: 23.662


In [43]:
# evaluate the model
error = model.evaluate(X_test, y_test, verbose=0)
print('MSE: %.3f, RMSE: %.3f' % (error, sqrt(error)))

# make second prediction
row = [0.17004,12.5,7.87,0,0.524,6.004,85.9,6.5921,5,311,15.2,386.71,17.1]
yhat = model.predict([row])
print('Predicted: %.3f' % yhat)

MSE: 48.764, RMSE: 6.983
Predicted: 18.443


In [44]:
# evaluate the model
error = model.evaluate(X_test, y_test, verbose=0)
print('MSE: %.3f, RMSE: %.3f' % (error, sqrt(error)))

# make third prediction
row = [0.22489,12.5,7.87,0,0.524,6.377,94.3,6.3467,5,311,15.2,392.52,20.45]
yhat = model.predict([row])
print('Predicted: %.3f' % yhat)

MSE: 48.764, RMSE: 6.983
Predicted: 16.985
